doc/decisions/0_drafts/errors_core.md

```diff
@@ -278,12 +287,77 @@ if (found == NULL)
 
 ## Decision
 
-**Suggestion:** Go with "Pragmatic solution"
+Go with "Pragmatic solution".
+We don't define any strict rules for the entire API.
+Instead, we define a few basic guidelines.
+
+### Guidelines
+
+These guidelines shall be applied to each function of the public API separately to decide how the function treats errors.
+
+#### Boolean functions
+
+If a function is conceptually boolean, it should return a `bool`.
+This means choosing `true` or `false` as the fallback value used in case of errors.
+
+If choosing `true` or `false` is not feasible, redesigning the function may be necessary.
+The function may be doing to many checks in one go and splitting it into multiple functions could make sense.
+
+If splitting is also not possible, an `int` or `enum` should be returned.
+The name of such a function must make clear that it is not actually boolean (avoid `is` and `has`, prefer `check`) to avoid accidental use in an `if`.
+
+#### Check preconditions
+
+Functions must check all preconditions.
+A (bug-free) function should never segfault, abort the process or otherwise cause unexpected behavior on the caller side.
+All conditions that could cause such behavior must be listed as preconditions and checked by the function as a safety measure.
+
+#### `NULL` arguments
+
+A `NULL` pointer argument should never be treated as a separate error case (unless it is the only error case).
+If a `NULL` pointer constitutes an invalid argument, the function should return an existing error (might be used for other precondition checks) or a default fallback value.
+
+#### Simple preconditions
+
+Simple preconditions are those that can be checked easily by the caller and which are not computationally expensive to check.
+All such preconditions should use the same error code, which may even be shared with other error cases.
+
+#### Complex preconditions
+
+Complex preconditions which
+
+- cannot (easily) be checked by the caller (e.g., because there is no public API for it) OR
+- are computationally expensive to check (i.e., duplicating the check would be bad)
+
+should always use separate error codes for each error case.
+The error code may not be shared with other complex preconditions, but can be shared with simple preconditions or `NULL` arguments, since those are easy to distinguish for the caller.
+
+#### Restructure API
+
+To avoid complex preconditions and the need for multiple error codes, the API can be restructured.
+This should, however, be done with caution.
+The API should still remain ergonomic to use and there should not be extra memory allocations creating opaque objects just to guard against possible errors.
```

Comment:
I agree this is what we both want to reach with this decision. But without details how it is done, the "rule" is worthless. (Imho it is not a rule but the goal of the decision.)

Reply:
I'm not sure what you want to say here.

This guideline is meant convey that splitting a function into multiple smaller functions with only one failure case each, **can** be an option. But in most cases to the API would only be safe to use, if additional opaque structs are introduced, which are then passed from one function to the next. If this structs aren't used internally somehow and are just temporary allocations, they shouldn't exist. It is better to have a single function returning an `int` with multiple failure cases and avoid the temporary allocations.